lightblue/src/main/java/org/esbtools/eventhandler/lightblue/LightblueDocumentEventRepository.java

```diff
@@ -0,0 +1,294 @@
+            try {
+                // TODO: Parameterize lock polling interval
+                // Or can we do lock call that only returns once lock is available?
+                Thread.sleep(3000L);
+            } catch (InterruptedException e) {
+                // Just try again...
+            }
+        }
+    }
+
+    /**
+     * Creates {@link LightblueDocumentEvent}s from entities using {@link #documentEventFactory},
+     * optimizes away superseded and merge-able events, and populates {@code entitiesToUpdate} with
+     * entity status updates that should be persisted to the document event entity collection
+     * before releasing locks.
+     *
+     * <p>The returned list may include net-new events as the result of merges. These events do not
+     * yet have an associated <em>persisted</em> entity, and therefore have no id's.
+     *
+     * @param maxEvents The maximum number of parsed document events to return. Not to be confused
+     *                  with {@link #documentEventBatchSize}.
+     * @param documentEventEntities The priority-first entities from lightblue.
+     * @param entitiesToUpdate A mutable collection of entities who's status modifications should be
+     *                         persisted back into lightblue.
+     * @param requester Used by {@link LightblueDocumentEvent} implementations to lookup their
+     *                  corresponding documents.
+     * @return An optimized list of parsed or merged document events ready to be published. No more
+     * than {@code maxEvents} will be returned. List may include events with newly computed entities
+     * that are not yet persisted.
+     */
+    private List<LightblueDocumentEvent> parseAndOptimizeDocumentEventEntitiesUpTo(int maxEvents,
+            DocumentEventEntity[] documentEventEntities, List<DocumentEventEntity> entitiesToUpdate,
+            LightblueRequester requester) {
+        List<LightblueDocumentEvent> optimized = new ArrayList<>(maxEvents);
+
+        for (DocumentEventEntity newEventEntity : documentEventEntities) {
+            LightblueDocumentEvent newEvent = documentEventFactory
+                    .getDocumentEventForEntity(newEventEntity, requester);
+
+            Iterator<LightblueDocumentEvent> optimizedIterator = optimized.iterator();
+
+            // We have a new event, let's see if it is superseded by or can be merged with any
+            // previous events we parsed or created as a result of a previous merge.
+            while (optimizedIterator.hasNext()) {
+                LightblueDocumentEvent previouslyOptimizedEvent = optimizedIterator.next();
+
+                if (newEvent.isSupersededBy(previouslyOptimizedEvent)) {
+                    // Keep previous event...
+                    DocumentEventEntity previousEntity = previouslyOptimizedEvent.wrappedDocumentEventEntity();
+                    previousEntity.addSurvivorOfIds(newEventEntity.getSurvivorOfIds());
```

Comment:
The new event may be one that was already created as a result of a merge, therefore it will have already survived some events. We want to make sure we track those as well.